fasteners/threads.scad

```diff
@@ -44,32 +59,140 @@ use <scad-utils/transformations.scad>
     diameter = 8,
     pitch = 1,
     length = 1,
     internal = false,
+    clearance = 0,
     n_starts = 1
 )
 {
+    // Tolerancing as per ISO 286
+    // See https://en.wikipedia.org/wiki/IT_Grade
+    function IT_grade (diameter, ITG) =
+        assert(ITG>=1, "ITG has to be between 1 and 16")
+        assert(ITG<=16, "ITG has to be between 1 and 16")
+        let (i = 0.45 * pow(diameter, 1/3) + 0.001 * diameter)
+            pow(10, 0.2*(ITG-1)) * i / 1000;
+
+    // Fundamental deviations as per ISO 286. Formulas obtained from:
+    // https://www.cobanengineering.com/Tolerances/ToleranceBand.asp
+    function EI_es_g (diameter) = 2.5 * pow(diameter, 0.34) / 1000;
+
+    theta = 60;                // V-shape angle, 60º for metric threads
+    H = pitch/2/tan(theta/2);  // height of the V-shape
+    tolerance = internal
+        ? IT_grade(diameter, 6)/2 + 0                  // H6
+        : IT_grade(diameter, 6)/2 + EI_es_g(diameter); // g6
+    D_maj = internal
+        ? diameter + tolerance + clearance
+        : diameter - tolerance - clearance;
+    D_min = D_maj - 2*5/8*H;
+
     trapezoidal_thread (
         pitch = pitch,
         length = length,
         upper_angle = 30, lower_angle = 30,
         outer_flat_length = pitch / 8,
-        major_radius = diameter / 2,
-        minor_radius = diameter / 2 - 5/8 * cos(30) * pitch,
+        major_radius = D_maj / 2,
+        minor_radius = D_min / 2,
         internal = internal,
         n_starts = n_starts
     );
 }
 
+/**
+ * Applies chamfers to a thread object. Expects 2 or 3 children.
+ *
+ * Children:
+ *   1) thread object
+ *   2) chamfer object at near end (Z = 0)
+ *   3) (optional) chamfer object at far end (Z = length)
+ * Parameters:
+ *   length   - axial length in mm of the thread object.
+ *   internal - true when the thread is internal, false when external.
+ */
+module chamfered_thread (length, internal = false)
+{
+    if (internal) {
+        union() {
+            children(0);
+            children(1);
+            if ($children == 3) {
+                translate([0,0,length]) rotate([180,0,0]) children(2);
+            }
+        }
+    }
+    else {
+        difference() {
+            children(0);
+            children(1);
+            if ($children == 3) {
+                translate([0,0,length]) rotate([180,0,0]) children(2);
+            }
+        }
+    }
+}
+
+/*
+ * Generates a revolved chamfer for a cylinder end face. For external
+ * cylinders this object can be subtracted from the cylinder. For
+ * internal cylinders, this object can be subtracted along with the
+ * cylinder from the original solid.
+ *
+ * cyl_diameter  - diameter of the cylinder to generate a chamfer for.
+ * chfr_diameter - max (min) diameter of the chamfer feature on internal
+ *                 (external) cylinders
+ * angle         - chamfer lead-in angle
+ * internal      - true when the cylinder is internal, false when external.
+ */
+module chamfer_cylinder (
```

Comment:
Regarding this module, requiring the user to calculate `chfr_diameter` seems like sub-optimal UX, especially if we're just going to fail assertions if the user passed in wrong values.

How about allowing the user to specify a `chamfer_depth` (`h` in the ascii-art diagram below) instead, and calculating the chamfer diameter from that and the other parameters?

The calculations could probably work like so:

```openscad
d = chamfer_depth * tan(angle);
chfr_diameter = (internal) ? (cyl_diameter + d) : (cyl_diameter - d);
```

Reply:
Good point this one, I ended up constantly calculating `chfr_diameter` myself. The reason behind it is that the chamfer diameter and the angle are usually what's defined in drawings since they are they two features easier to inspect. Anyway, I'm happy to change that, I agree the UX wasn't the best. However, I'd rather use `d` (as defined in the diagram) and the cylinder diameter as the parameters. Particularly on internal chamfers, it is very rare to define them as an angle and a depth. That depth could potentially be impossible to measure. So how about going with `diameter`as in (cylinder diameter) and `distance` (which equals to `d` in the diagram and is the name used on other CAD programs). I'm easy with the naming anyway. Let me know what you think.

Reply:
Hmm, interesting. After reading the above, I looked up chamfer dimensioning and found http://draftingmanuals.tpub.com/14276/css/Chamfers-163.htm.

Looks like both `depth` and `chamfer_diameter` options are used (the former in the external chamfer diagram, and the latter in the internal chamfer diagram.

So how about allowing both options then? Set both parameters to `undef` by default, and calculate the missing one.

As for naming it `distance`, I think that seems as bit vague (in the case of non-45°-chamfers, which distance? the horizontal one or the vertical one?)